be/src/storage/lake/local_pk_index_manager.cpp

```diff
@@ -0,0 +1,163 @@
+        auto pk_path = data_dir->get_persistent_index_path();
+        LOG(INFO) << "start to gc local persistent index dir:" << pk_path;
+        for (const auto& tablet_id : tablet_ids) {
+            auto tablet_pk_path = pk_path + "/" + tablet_id;
+            int64_t id = std::stoll(tablet_id);
+            // judge whether tablet should be in the data_dir or not,
+            // for data_dir may change if config:storage_path changed.
+            // just remove if not.
+            if (StorageEngine::instance()->get_persistent_index_store(id) != data_dir) {
+                dir_changed_tablet_ids.push_back(id);
+                if (StorageEngine::instance()->clear_persistent_index(data_dir, id, tablet_pk_path).ok()) {
+                    removed_dir_tablet_ids.push_back(id);
+                }
+            } else if (!tablet_manager->is_tablet_in_worker(id)) {
+                // the shard may be scheduled to other nodes
+                if (!update_manager->try_lock_pk_index_shard(id)) {
+                    LOG(WARNING) << "Fail to lock pk index, tablet id: " << id;
+                    continue;
+                }
+                not_in_worker_tablet_ids.emplace_back(id);
+                // judge whether tablet is scheduled again,
+                // and pk_index_shard write_lock has been hold, so no process will build the persistent index.
+                if (!tablet_manager->is_tablet_in_worker(id)) {
+                    // try to remove pk index cache to avoid continuing to use the index in the cache after deletion.
+                    if (update_manager->try_remove_primary_index_cache(id)) {
+                        if (StorageEngine::instance()->clear_persistent_index(data_dir, id, tablet_pk_path).ok()) {
+                            removed_dir_tablet_ids.push_back(id);
+                        }
+                    }
+                }
+                update_manager->unlock_pk_index_shard(id);
+            }
+        }
+        int64_t t_end = MonotonicMillis();
+        LOG(INFO) << "finish gc local persistent index dir: " << pk_path
+                  << ", found tablet not in the worker, tablet_ids: " << JoinInts(not_in_worker_tablet_ids, ",")
+                  << ", data_dir changed tablet_ids: " << JoinInts(dir_changed_tablet_ids, ",")
+                  << ", and removed dir successfully, tablet_ids: " << JoinInts(removed_dir_tablet_ids, ",")
+                  << ", cost:" << t_end - t_start << "ms";
+    }
+}
+
+bool LocalPkIndexManager::need_evict_tablet(const std::string& tablet_pk_path) {
+    auto now = time(nullptr);
+    auto mtime_or = FileSystem::Default()->get_file_modified_time(tablet_pk_path);
+    if (!mtime_or.ok()) {
+        return false;
+    }
+    auto mtime = *mtime_or;
+    if (now - mtime < config::lake_local_pk_index_unused_threshold_seconds) {
+        return false;
+    }
+
+    std::set<std::string> index_files;
+    Status st = fs::list_dirs_files(tablet_pk_path, nullptr, &index_files);
```

Comment:
It seems that `FileSystem::iterate_dir2` is more suitable here? It can reduce memory usage and reduce unnecessary directory entry fetch.